server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/SearchAction.java

```diff
@@ -0,0 +1,159 @@
+    String projectUuid = request.param(PARAM_PROJECT_ID);
+    String projectKey = request.param(PARAM_PROJECT_KEY);
+    List<String> fieldsToReturn = request.paramAsStrings(WebService.Param.FIELDS);
+    SearchOptions searchOptions = new SearchOptions()
+      .setPage(request.mandatoryParamAsInt(WebService.Param.PAGE),
+        request.mandatoryParamAsInt(WebService.Param.PAGE_SIZE));
+
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      ComponentDto project = searchProject(dbSession, projectUuid, projectKey);
+      List<CustomMeasureDto> customMeasures = dbClient.customMeasureDao().selectByComponentUuid(dbSession, project.uuid(), searchOptions);
+      int nbCustomMeasures = dbClient.customMeasureDao().countByComponentUuid(dbSession, project.uuid());
+      Map<String, User> usersByLogin = usersByLogin(customMeasures);
+      Map<Integer, MetricDto> metricsById = metricsById(dbSession, customMeasures);
+      JsonWriter json = response.newJsonWriter();
+      json.beginObject();
+      customMeasureJsonWriter.write(json, customMeasures, project, metricsById, usersByLogin, fieldsToReturn);
+      searchOptions.writeJson(json, nbCustomMeasures);
+      json.endObject();
+      json.close();
+    } finally {
+      MyBatis.closeQuietly(dbSession);
+    }
+  }
+
+  private Map<Integer, MetricDto> metricsById(DbSession dbSession, List<CustomMeasureDto> customMeasures) {
+    List<MetricDto> metrics = dbClient.metricDao().selectByIds(dbSession, Lists.transform(customMeasures, new CustomMeasureToMetricIdFunction()));
+    return Maps.uniqueIndex(metrics, new MetricToIdFunction());
+  }
+
+  private Map<String, User> usersByLogin(List<CustomMeasureDto> customMeasures) {
+    return FluentIterable.from(customMeasures)
+      .transform(new CustomMeasureToUserLoginFunction())
+      .toMap(new UserLoginToUserFunction());
+  }
+
+  private ComponentDto searchProject(DbSession dbSession, @Nullable String projectUuid, @Nullable String projectKey) {
+    checkArgument(projectUuid != null ^ projectKey != null, "The project id or project key must be provided, not both.");
+    ComponentDto project;
+    if (projectUuid != null) {
+      project = dbClient.componentDao().selectByUuid(dbSession, projectUuid);
+    } else {
+      project = dbClient.componentDao().selectByKey(dbSession, projectKey);
+    }
+
+    return project;
+  }
+
+  private static final class CustomMeasureToUserLoginFunction implements Function<CustomMeasureDto, String> {
+    @Override
+    public String apply(@Nonnull CustomMeasureDto customMeasure) {
+      return customMeasure.getUserLogin();
+    }
+  }
+
+  private final class UserLoginToUserFunction implements Function<String, User> {
+    @Override
+    public User apply(@Nonnull String userLogin) {
+      return userIndex.getByLogin(userLogin);
+    }
+  }
+
+  private final class CustomMeasureToMetricIdFunction implements Function<CustomMeasureDto, Integer> {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this a "static" inner class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2694)